### PR TITLE
Matrix4x4.CreateViewport

### LINKED ISSUE #79961: [API Proposal]: Matrix4x4.CreateViewport
### Background and motivation

`Matrix4x4` currently provides a large number of methods for creating various transformation matrices. The figure below shows a geometric pipeline for transforming vertices from object coordinates to window coordinates. Unfortunately, `Matrix4x4` does not provide a method for creating a matrix that transforms vertices from normalized device coordinates to window coordinates.

image

I propose to add a method for creating a viewport matrix, which is shown in the figure below. `width` and `height` specify viewport dimensions, `x` and `y` stand for coordinates of the upper left corner of the viewport, `zmin` and `zmax` define the minimum and maximum viewport depth.

image

### API Proposal

```csharp
namespace System.Numerics;

public struct Matrix4x4
{
    public static Matrix4x4 CreateViewport(float x, float y, float width, float height, float minDepth, float maxDepth)
    {
        Matrix4x4 result = Identity;
        
        result.M11 = width * 0.5f;
        result.M22 = -height * 0.5f;
        result.M33 = maxDepth - minDepth;
        result.M41 = x + result.M11;
        result.M42 = y - result.M21;
        result.M43 = minDepth;

        return result;
    }
}
```

### API Usage

```csharp
Matrix4x4 scaleMatrix = Matrix4x4.CreateScale(model.Scale);
Matrix4x4 rotationMatrix = Matrix4x4.CreateFromYawPitchRoll(model.Yaw, model.Pitch, model.Roll);
Matrix4x4 translationMatrix = Matrix4x4.CreateTranslation(model.Position);
Matrix4x4 modelMatrix = scaleMatrix * rotationMatrix * translationMatrix;
Matrix4x4 viewMatrix = Matrix4x4.CreateLookAt(camera.Position, camera.Target, camera.Up);
Matrix4x4 projectionMatrix = Matrix4x4.CreatePerspectiveFieldOfView(camera.Fov, camera.AspectRatio, camera.Near, camera.Far);
Matrix4x4 modelViewProjectionMatrix = modelMatrix * viewMatrix * projectionMatrix;
Matrix4x4 viewportMatrix = Matrix4x4.CreateViewport(0, 0, bitmap.PixelWidth, bitmap.PixelHeight, 0, 1);

Vector4[] windowVertices = new Vector4[model.Vertices.Length];
for (int i = 0; i < windowVertices.Length; i++)
{
    windowVertices[i] = Vector4.Transform(model.Vertices[i], modelViewProjectionMatrix);
    windowVertices[i] /= windowVertices[i].W;
    windowVertices[i] = Vector4.Transform(windowVertices[i], viewportMatrix);
}

DrawPoints(windowVertices);
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by AlexRadch**
Close https://github.com/dotnet/runtime/issues/79961 issue.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.Impl.cs [hunk 1]
```diff
@@ -832,6 +832,22 @@ public static Impl CreateTranslation(float positionX, float positionY, float pos
...
+
+                result.X.X = width * 0.5f;
+                result.Y.Y = -height * 0.5f;
+                result.Z.Z = maxDepth - minDepth;
+                result.W.X = x + result.X.X;
+                result.W.Y = y - result.Y.Y;
+                result.W.Z = minDepth;
+
+                return result;
+            }
```
[the remark this answers is not in the document]
Comment (AlexRadch) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.Impl.cs]: Your suggestion has a small issue. I fixed it.

**Comment by AlexRadch** [comment]
Are there ways to fast build `System.Private.CoreLib.Shared.shproj` project?
 Now I run `build ` from the root folder.
 It takes 15 minutes or more to build 😥

**Review by tannergooding — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.Impl.cs [hunk 2]
```diff
@@ -832,6 +832,24 @@ public static Impl CreateTranslation(float positionX, float positionY, float pos
...
 
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Impl CreateViewport(float x, float y, float width, float height, float minDepth, float maxDepth)
+            {
+                Impl result;
+
+                // 4x SIMD fields to get a lot better codegen
+                result.W = new Vector4(width, height, 0f, 0f);
+                result.W *= new Vector4(0.5f, 0.5f, 0f, 0f);
```
Comment (IDisposable) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.Impl.cs:843]: Couldn't this be `result.W = new Vector4(width * 0.5f, height * 0.f, 0f, 0f);`?

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.Impl.cs:843]: That's less efficient as it requires two separate multiplications and then an insert.

The way its done here is 1 insert and then 1 multiplication.

**Pull request merged**